//
// Bentobox
// Proto
// Services
//

syntax = "proto3";
package bento.protos;

import "bento/protos/values.proto";
import "bento/protos/references.proto";
import "bento/protos/sim.proto";

service EngineService {
  // Get version information
  rpc GetVersion(GetVersionReq) returns (GetVersionResp);

  // Create, Update, Get, List and Delete Simulations
  rpc ApplySimulation(ApplySimulationReq) returns (ApplySimulationResp);
  rpc GetSimulation(GetSimulationReq) returns (GetSimulationResp);
  rpc ListSimulation(ListSimulationReq) returns (ListSimulationResp);
  rpc DropSimulation(DropSimulationReq) returns (DropSimulationResp);

  // Run one step of the simulation
  rpc StepSimulation(StepSimulationReq) returns (StepSimulationResp);

  // Set, Get values in components
  rpc GetValue(GetValueReq) returns (GetValueResp);
  rpc SetValue(SetValueReq) returns (SetValueResp);
}

message GetVersionReq {}

message GetVersionResp {
  // VCS Commit Hash which the Engine running the simulation service is built
  // on.
  string commit_hash = 1;
}

message ApplySimulationReq {
  // Definition of the simulation to create or apply
  SimulationDef simulation = 1;
}
message ApplySimulationResp {
  // Simulation def with all unspecified entity and system IDs created
  SimulationDef simulation = 1;
}

message GetSimulationReq {
  // Name of the simulation to get
  string name = 1;
}
message GetSimulationResp {
  SimulationDef simulation = 1;
}

message ListSimulationReq {}
message ListSimulationResp {
  // list of names representing all the currently registered simulations
  repeated string sim_names = 1;
}

message DropSimulationReq {
  // Name of the simulation to drop
  string name = 1;
}
message DropSimulationResp {}

message StepSimulationReq {
  // Name of the simulation to step
  string name = 1;
}
message StepSimulationResp {}

message GetValueReq {
  // Name of the simulation to retrieve the value from
  string sim_name = 1;
  // Reference to the value to retrieve
  ValueRef value_ref = 2;
}
message GetValueResp {
  // Stored value of the requested value
  Value value = 1;
}

message SetValueReq {
  // Name of the simulation to set the value
  string sim_name = 1;
  // Reference to the target value to set
  ValueRef value_ref = 2;
  // The value to the set the target ValueRef to
  Value value = 3;
}
message SetValueResp {}
